src/histbuf.rs

```diff
@@ -0,0 +1,325 @@
+/// initial value is either the default value for the element type or a supplied
+/// initial value. This simplifies the API and is mostly irrelevant for the
+/// intended use case.
+///
+/// # Examples
+/// ```
+/// use heapless::HistoryBuffer;
+/// use heapless::consts::*;
+///
+/// // Initialize a new buffer with 8 elements, all initially zero.
+/// let mut buf = HistoryBuffer::<_, U8>::new();
+///
+/// buf.push(3);
+/// buf.push(5);
+/// buf.extend(&[4, 4]);
+///
+/// // The first (oldest) element is still zero.
+/// assert_eq!(buf.first(), &0);
+/// // The last (newest) element is a four.
+/// assert_eq!(buf.last(), &4);
+/// for el in buf.iter() { println!("{:?}", el); }
+///
+/// // Now we can prepare an average of all values, which comes out to 2.
+/// let avg = buf.iter().sum::<usize>() / buf.len();
+/// assert_eq!(avg, 2);
+/// ```
+#[derive(Clone)]
+pub struct HistoryBuffer<T, N>
+where
+    N: ArrayLength<T>,
+{
+    data: GenericArray<T, N>,
+    write_at: usize,
+}
+
+
+impl<T, N> HistoryBuffer<T, N>
+where
+    N: ArrayLength<T>,
+    T: Default,
+{
+    /// Constructs a new history buffer, where every element is filled with the
+    /// default value of the type `T`.
+    ///
+    /// `HistoryBuffer` currently cannot be constructed in `const` context.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use heapless::HistoryBuffer;
+    /// use heapless::consts::*;
+    ///
+    /// // Allocate a 16-element buffer on the stack
+    /// let mut x: HistoryBuffer<u8, U16> = HistoryBuffer::new();
+    /// // All elements are zero
+    /// assert!(x.iter().eq([0; 16].iter()));
+    /// ```
+    pub fn new() -> Self {
+        Self {
+            data: Default::default(),
+            write_at: 0,
+        }
+    }
+
+    /// Clears the buffer, replacing every element with the default value of
+    /// type `T`.
+    pub fn clear(&mut self) {
+        *self = Self::new();
+    }
+}
+
+impl<T, N> HistoryBuffer<T, N>
+where
+    N: ArrayLength<T>,
+    T: Clone,
+{
+    /// Constructs a new history buffer, where every element is the given value.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use heapless::HistoryBuffer;
+    /// use heapless::consts::*;
+    ///
+    /// // Allocate a 16-element buffer on the stack
+    /// let mut x: HistoryBuffer<u8, U16> = HistoryBuffer::new_with(4);
+    /// // All elements are four
+    /// assert!(x.iter().eq([4; 16].iter()));
+    /// ```
+    pub fn new_with(t: T) -> Self {
+        Self {
+            data: GenericArray::generate(|_| t.clone()),
+            write_at: 0,
+        }
+    }
+
+    /// Clears the buffer, replacing every element with the given value.
+    pub fn clear_with(&mut self, t: T) {
+        *self = Self::new_with(t);
+    }
+}
+
+impl<T, N> HistoryBuffer<T, N>
+where
+    N: ArrayLength<T>,
+{
+    /// Returns the length of the buffer, which is the length of the underlying
+    /// backing array.
+    pub fn len(&self) -> usize {
```

Comment:
Q: is there any use case for this returning not the capacity but the number of elements pushed in the buffer after clearing it (or starting from a cleared buffer)? As in `HIstoryBuffer::new().len()` should return 0.

Reply:
Well, my use case is a buffer that once created is filled continuously with new values, so the "len < capacity" state is only observed for a very short time.

With the semantics that `len()` can be smaller than capacity, you also need to have basically full Vec-like semantics, i.e. `first()` and `last()` returning option, which I wanted to avoid. But it may be cleaner to do it that way.

(In that case, the implementation can probably use a `Vec` internally.)